Use fast integer division algorithm to avoid division ops inside kernels.

yongjik opened the pull request:
- OffsetInfo and OffsetIterator pre-computes the necessary coordinate
  change along each dimension, so that each successive offset can be
  computed using only addition/subtraction/comparisons.

- Added IntDivider which supports "magic division" for uint32_t, thus
  eliminating integer divisions altogether for offset calculation, as
  long as indices fit in 32 bits.

- In code paths with statically determined dimensions (Dims=1 or 2),
  kernel arguments now contain only the necessary data (instead of
  MAX_CUTORCH_DIMS of everything).

- Fixed index overflow errors: for tensors with >= 2G elements, we used
  to have incorrect results or an infinite loop inside the kernel.

TODO: The following pattern is broken for tensors with >= 2G elements.
      It will result in overflow, even if IndexType is uint64_t.  Need
      to search and replace them.

  > for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
  >      linearIndex < totalElements;
  >      linearIndex += gridDim.x * blockDim.x) {

yongjik committed:
Use pre-computed offset increments to avoid int division inside kernels.

- OffsetInfo and OffsetIterator pre-computes the necessary coordinate
  change along each dimension, so that each successive offset can be
  computed using only addition/subtraction/comparisons.

- Added IntDivider which supports "magic division" for uint32_t, thus
  eliminating integer divisions altogether for offset calculation, as
  long as indices fit in 32 bits.

- In code paths with statically determined dimensions (Dims=1 or 2),
  kernel arguments now contain only the necessary data (instead of
  MAX_CUTORCH_DIMS of everything).

- Fixed index overflow errors: for tensors with >= 2G elements, we used
  to have incorrect results or an infinite loop inside the kernel.

TODO: The following pattern is broken for tensors with >= 2G elements.
      It will result in overflow, even if IndexType is uint64_t.  Need
      to search and replace them.

  > for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
  >      linearIndex < totalElements;
  >      linearIndex += gridDim.x * blockDim.x) {

yongjik commented:
This PR improves some float operations by ~20% (and some operations on ByteTensor by up to ~45%), but in general the performance impact seems small, unless one uses a lot of non-contiguous tensors and/or broadcasting with large dimensions.

Here's an example where I could get ~20% improvement on GTX 1080:

    A = torch.cuda.FloatTensor(1000, 256)
    B = torch.cuda.FloatTensor(128)
    A = A[:, :128]
    A.pow_(B)  # Improves from ~7.6 to ~6.0 usec

I found at least one case where it becomes slower by ~5%, but such cases seem to be rare, so I still think it's a net performance win on average, although small.

    A = torch.cuda.IntTensor(2048, 2048)
    A = A[:, :2000]
    B = torch.cuda.IntTensor(2000).fill_(10)
    A.remainder_(B)  # Changes from ~180 to 189 usec.

Raw benchmark results are https://github.com/yongjik/pt_test/tree/master/results/offset in case anybody's interested.

* This PR also fixes overflow errors with large tensors (2G elements or more), which makes me suspect that nobody has actually been using 64-bit index math so far...

yongjik committed:
Merge branch 'master' into 0205-offset

yongjik committed:
Update CMakeLists.txt

soumith commented:
@pytorchbot add to whitelist

wickedfoo commented on aten/src/THC/THCOffsetInfo.cuh:
```diff
@@ -0,0 +1,274 @@
+  bool carry = false;
+
+  for (int i = dims - 1; i > 0; --i) {
+    IndexType index = indices[i] + increments[i] + (IndexType) carry;
```
I'm not convinced that maintaining the increment this way is faster than it was before, or especially if you simplify the integer div/mod as above and just calculate the index each time.

This is a lot of instructions to go through, plus the predicate at line 81. I have to imagine that this will bloat out the register count as well, as you're maintaining per-dimension increment state.

The kernels are not as easy to read as before either.

Can you try this code without adding these iterator operators, and have it as before, just by modifying the linear index -> offset code?

yongjik replied:
Done.

wickedfoo commented on aten/src/THC/THCApply.cuh:
```diff
@@ -119,17 +120,18 @@ template <typename Op,
 __launch_bounds__(THC_APPLY_THREADS_PER_BLOCK, THC_APPLY_BLOCKS_PER_SM)
```
these bounds can probably just be removed, they were a pretty old hack to get better occupancy on CC 3.5 (K40 generation).

yongjik replied:
Done.

wickedfoo commented on aten/src/THC/THCOffsetInfo.cuh:
```diff
@@ -0,0 +1,274 @@
+    IndexType next = index + step;
+
+    // The second condition is necessary to handle overflow (e.g., when step is
+    // 2GB and limit is 3GB, assuming 32-bit index).
```
In the case where the step is 2 GB and the limit is 3 GB, why is the index 32 bits?

If this is the case, this feels like a bug in the code used to determine whether or not 64 bit indexing math is needed. This particular check shouldn't be needed, as it should be guarded against before we even get here.

yongjik replied:
You can see the function here:

https://github.com/pytorch/pytorch/blob/78649419c42c157db1c624df5e5020a78b007dd6/aten/src/THC/THCTensorTypeUtils.cuh#L75

https://github.com/pytorch/pytorch/blob/78649419c42c157db1c624df5e5020a78b007dd6/aten/src/THC/THCTensorTypeUtils.cu#L247

It basically admits any tensor with size up to `UINT32_MAX - 1`.  Shall I change it to instead accept up to `INT32MAX - 1` ?  (On one hand, tensors with 2-4G elements will now have to use 64-bit math.  On the other hand, given that the code basically didn't work before, I think nobody will be affected by the slowdown.)

wickedfoo replied:
Changing it to INT32MAX - 1 is fine, people would hardly use such large sizes and this is just to make sure it still executes correctly.

yongjik replied:
Done.

wickedfoo commented on aten/src/THC/THCApply.cuh:
```diff
@@ -141,22 +143,21 @@ template <typename Op,
+                      const OffsetInfo<Tb, IndexType, BDims> b,
                       IndexType totalElements,
                       Op op) {
-  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
```
keeping these old for loops, as you found out, cast blockIdx.x to IndexType etc. to avoid overflow

yongjik replied:
Done.

wickedfoo reviewed (changes requested):
I'm not convinced that the code as you have it results in a performance win, and it makes the code a lot more complicated. 7.6 us to 6.0 us is within the realm of noise, and such changes are sensitive to heuristics used in the register allocator and in other places.

Replacing the linear index with a per-dimension index will bloat out the register count, and the code within the new iteration stuff looks like it has divergent/predicated execution paths as well.

However, I do believe that constant integer division via multiplication/shift by constants is worth trying. Your magic number division algorithm can be simplified by restricting its usage to the case 2 to max signed int (see comments).

Can you do a more minimal diff keeping the old kernel structure and the linear index -> offset lookup trying the faster version of the magic constant division algorithm, with a fallback to using normal integer div/mod if it falls outside the range under consideration?

For performance testing, I would concentrate on sufficiently large tensor sizes, say a large tensor (multi-100 MB+ in size) that is transposed on which you perform pointwise operations. A kernel that executes in just microseconds I think is likely to fall within the margin of noise.

Also I would inspect the SASS to see what instructions it was emitting before for integer div/mod (I believe it tries to map it to floating point inverse, when I recall looking a long time ago), and see what instructions it actually issues for umulhi as well.

yongjik commented:
Hi @wickedfoo, thanks for the detailed review, and I understand your point that the code is too complicated for the (rather unimpressive) speedup.  I'll try just using the constant division algorithm and get back to you.  Might take a few days.

On the other hand, I do think there's a measurable speedup for some cases.  One case I found:

    A = torch.cuda.FloatTensor(4096, 4096)
    B = torch.cuda.FloatTensor(2048)
    A = A[:, :2048]
    A.pow_(B)  # Changes from ~389 us to ~315 us (speedup ~23%)

Ironically, using even larger tensor doesn't show larger speedup, because then (I suppose) memory bandwidth dominates everything.

ngimel commented:
@yongjik also take a look at https://github.com/milakov/int_fastdiv

yongjik commented on aten/src/THC/THCIntegerDivider.cuh:
```diff
@@ -0,0 +1,120 @@
+#ifdef __CUDA_ARCH__
+    // 't' is the higher 32-bits of unsigned 32-bit multiplication of 'n' and
+    // 'm1'.
+    unsigned int t = __umulhi(n, m1);
```
I just built two toy kernels to compare, and the relevant SASS code for sm_61 (GTX 1080) looks like this:

    // *data = *data * a;

        ...
        /*0028*/ LDG.E R0, [R2];
        /*0030*/ XMAD R5, R0.reuse, c[0x0] [0x148], RZ;
        /*0038*/ XMAD.MRG R6, R0.reuse, c[0x0] [0x148].H1, RZ;

        /*0048*/ XMAD.PSL.CBCC R0, R0.H1, R6.H1, R5;
        /*0050*/ STG.E [R2], R0;
        ...

    // *data = __umulhi(*data, a);

        ...
        /*0028*/ LDG.E R0, [R2];
        /*0030*/ XMAD R5, R0.reuse, c[0x0] [0x148], RZ;
        /*0038*/ XMAD R4, R0.reuse, c[0x0] [0x148].H1, RZ;

        /*0048*/ XMAD R6, R0.H1, c[0x0] [0x148].H1, RZ;
        /*0050*/ XMAD.CHI R5, R0.H1, c[0x0] [0x148], R5;
        /*0058*/ IADD3.RS R0, R5, R4, R6;

        /*0068*/ STG.E [R2], R0;
        ...

XMAD is listed as "Integer Short Multiply Add", so it looks like the difference is between (3 FMA operations) vs. (4 FMA operations + 1 addition): it looks reasonable to me, unless .CHI is some secret codeword for "super heavy slow computation".

(While we're at it, is there any public documentation for SASS instructions?  I could only find http://docs.nvidia.com/cuda/cuda-binary-utilities/index.html which is not much help.)

wickedfoo replied:
There's no public documentation for it (it also varies across architectures), though some people have reverse engineered it. Inspecting the code that it generates is useful to determine if something really bad is happening (emulated instructions, etc.)

This seems reasonable to me.

Out of curiosity, can you try signed 32-bit integer multiplication and see what SASS it generates?

yongjik replied:
(Sorry for the late reply: have been busy with other stuff.  I'll upload the updated code by tonight.)

So I checked signed versions, and signed 32-bit multiplication is identical to unsigned version.

(It threw me off at first, but when you think about it, the lower 32 bits are the same for signed/unsigned 32-bit multiplications.)

Higher 32-bits for signed multiplication:

    // *data = __mulhi(*data, a);

        ...
        /*0028*/ LDG.E R0, [R2];
        /*0030*/ XMAD R5, R0.reuse, c[0x0] [0x148], RZ;
        /*0038*/ XMAD.U16.S16 R4, R0.reuse, c[0x0] [0x148].H1, RZ;

        /*0048*/ XMAD.S16.S16.CSFU R6, R0.H1, c[0x0] [0x148].H1, RZ;
        /*0050*/ XMAD.S16.U16.CHI R5, R0.H1, c[0x0] [0x148], R5;
        /*0058*/ IADD3.RS R0, R5, R4, R6;

        /*0068*/ STG.E [R2], R0;
        ...

Still 3 XMAD (fused multiply-add) ops and one IADD3, so it's basically the same as the unsigned version.

Just for fun, I tried signed/unsigned divisions, and they have:

    unsigned division: 14 XMAD ops, 8 IADD/IADD3/IADD32I ops, a few others.
    signed division: 14 XMAD ops, 9 IADD/IADD3/IADD32I ops, a few others.

wickedfoo commented:
Also the reason that IndexType was `unsigned int` rather than signed `int` for 32-bit indices was that I discovered that div/mod was faster on unsigned rather than signed types (relevant for K40 and M40 I think?), but this may no longer be the case with more recent architectures. It might be worth comparing both signed and unsigned int32 for IndexType to see if there's still a difference.

wickedfoo commented:
The integer division by magic constants code in the Caffe2 source I think will be faster than int_fastdiv if you exclude the -1 / 1 case. They're basically the same code more or less, except you avoid this additional work:

https://github.com/milakov/int_fastdiv/blob/master/int_fastdiv.h#L126

Stonesjtu commented:
@yongjik  I suffered a lot tuning the `indexSelectKernel` and I found the kernel is mainly limited by `read_throughput`. I suspects that we should focus more on reducing the memory read operations and improving the locality, rather than only computational cost.

yongjik committed:
Removed OffsetIterator, and kept only the fast integer division logic.

- Also changed canUse32BitIndexMath so that the max index for 32-bit
  math is INT32_MAX, instead of UINT32_MAX.  It also simplifies the
  division operation.

yongjik commented on aten/src/THC/THCIntegerDivider.cuh:
```diff
@@ -0,0 +1,120 @@
+    // 't' is the higher 32-bits of unsigned 32-bit multiplication of 'n' and
+    // 'm1'.
+    unsigned int t = __umulhi(n, m1);
+    unsigned int t2 = t + ((n - t) >> s1);
```
I changed `canUse32BitIndexMath` so that dividend/divisor are at most INT32_MAX.  Given the reduced range, the code can now work with `__umulhi`, one addition, and one shift.

(The Caffe2 code gets rid of the addition, but I think that requires 64-bit math.  An addition is probably better than adding a conditional check for `divisor==1`.)

yongjik commented on aten/src/THC/THCOffsetInfo.cuh:
```diff
@@ -0,0 +1,274 @@
+{
+  bool carry = false;
+
+  for (int i = dims - 1; i > 0; --i) {
```
The code is removed, so it doesn't matter, but it used to be only called when `dims` was a template parameter, so nvcc would unroll the loop fully.

yongjik commented:
Hi @wickedfoo, I updated the code to remove the increment stuff and only leave the int division algorithm.  Could you take another look?

Regarding signed/unsigned integer, I think the point is moot, because (in the references I found) the fast division algorithm for signed integers always has more operations than the unsigned version.  So I think they don't really give us any benefit here.

yongjik commented:
Hi guys, any thoughts on this PR?

yongjik commented:
Hi @wickedfoo, could you give your opinion?  If this PR still looks like too much complication, I understand if you don't want to merge this, but I'd appreciate a decision rather than this PR staying in limbo forever.  Thanks!

soumith commented:
@yongjik i think he does not get github notification emails. I will ping him directly. sorry for delay.

wickedfoo commented:
Looking now.

wickedfoo commented on aten/src/THC/THCOffsetInfo.cuh:
```diff
@@ -0,0 +1,89 @@
+#ifndef THC_OFFSET_INFO_INC
```
Can you move this into TensorInfo.cuh so it exists with the other offset code?

yongjik replied:
Done.

wickedfoo reviewed (commented):
Looks good to me. Any idea what the performance change of this is (i.e., is it worth it, and for what sizes)?

yongjik committed:
Merged OffsetInfo into THCTensorInfo.cuh.

yongjik commented on aten/src/THC/THCOffsetInfo.cuh:
```diff
@@ -0,0 +1,89 @@
+  __host__ __device__ T* get(IndexType linearIndex) const {
+    IndexType offset = 0;
+
+    for (int i = tinfo.dims - 1; i > 0; --i) {
```
Done: also changed IndexToOffset so that it skips the remainder operation for dim 0 (matches the code when Dims is statically known).

yongjik commented:
Hi @wickedfoo, thanks for the review.

I ran several hundred configurations of tensor operations (on GTX 1080 / CUDA 9.1), including `add`, `mul`, `tanh`, `pow`, and `remainder`, on tensors with sizes ranging between 1000x64 and 4096x4096.  About 1/3 of them showed speedup of 5% or more.  There are some cases when performance improves by up to ~50%.  For the rest, performance didn't change meaningfully (within 5%).  I saw several cases of slowdown of ~5%, but these operations were too small (< 20 us) and I couldn't reproduce them reliably: might be just random noise.

The biggest win I could find was:

    A = torch.cuda.HalfTensor(4096, 4096)
    B = torch.cuda.HalfTensor(2048)
    C = torch.cuda.HalfTensor(4096, 4096)

    # ~266 us to ~172 us (speedup 55%)
    torch.mul(A[:, :2048], B, out=C[:, :2048])

    A = torch.cuda.HalfTensor(8192, 8192)
    B = torch.cuda.HalfTensor(4096)
    C = torch.cuda.HalfTensor(8192, 8192)

    # ~1.06 to ~0.69 ms (speedup 54%)
    torch.mul(A[:, :4096], B, out=C[:, :4096])

We also have speedup for float operations, though not as dramatic:

    # ~26 to ~21 us (speedup 24%, but it's too noisy so might not be accurate)
    nrows, ncols, ncols2 = 1024, 1024, 512

    # ~365 to 312 us (speedup 17%)
    nrows, ncols, ncols2 = 4096, 4096, 2048

    # ~1.38 to 1.17 ms (speedup 18%)
    nrows, ncols, ncols2 = 8000, 8000, 4000

    A = torch.cuda.FloatTensor(nrows, ncols)
    B = torch.cuda.FloatTensor(ncols2)
    A = A[:, :ncols2]

    A.pow_(B)

For some other float operations, I observed speedup of ~25% for mid-size tensors (around 1000x128), but it becomes smaller as tensors get bigger (~9% for 1024x1024, ~3% for 8000x3000), probably because memory latency dominates everything for these tensors.

soumith commented:
@yongjik thanks a lot for the PR it's good to go.
@ezyang can you check if the failing perf-regression test is legit or false-positive?

yongjik commented:
I don't know if I'm doing it right, but I followed the advice of the failed test log and ran `cd .jenkins/perf_test/ && bash test_gpu_speed_mlstm.sh`:

On clean branch (2726550, ran three times):

    Runtime stats in seconds:
    {"mean": 2.3622500000000004, "sigma": 0.01499624953113275}
    {"mean": 2.37195, "sigma": 0.01663272376972571}
    {"mean": 2.3788499999999995, "sigma": 0.01288128487380043}

With this PR on top of it:

    {"mean": 2.3658499999999996, "sigma": 0.018205150370156273}
    {"mean": 2.3644000000000007, "sigma": 0.021511392330576876}
    {"mean": 2.36015, "sigma": 0.01142048597915163}

So I think there's no meaningful difference on GTX-1080, but other GPUs might report different numbers, I guess.

ezyang commented:
The GPU perf tests have been flaky recently, so you should ignore them for the purposes of assessing this PR.

soumith commented:
thanks @yongjik. sorry for the delay in review.

yongjik commented:
No worries! Half of the delay was mine, after all. Thanks for the review.
